### PR TITLE
QGIS Development Server Application

### PR DESCRIPTION
**Pull request by elpaso**
## QGIS Development HTTP Server

This is a small independent command line application that implements a minimal HTTP development server for QGIS Server.

It can be useful when developing and testing QGIS Server projects, modules and plugins without requiring a full web server/FCGI stack.

qgis_mapserver

The application is completely independent it does not introduce any change in the existing QGIS code, for this reason I'd like to merge it for 3.12 even if we are in feature freeze.

```bash
Usage: qgis_mapserver [options] [address:port]
QGIS Development Server

Options:
  -h, --help        Displays this help.
  -v, --version     Displays version information.
  -l <logLevel>     Sets log level (default: 0)
                    0: INFO
                    1: WARNING
                    2: CRITICAL
  -p <projectPath>  Path to a QGIS project file (*.qgs or *.qgz),
                    if specified it will override the query string MAP argument
                    and the QGIS_PROJECT_FILE environment variable

Arguments:
  addressAndPort    Listen to address and port (default: "localhost:8000")
                    address and port can also be specified with the environment
                    variables QGIS_SERVER_ADDRESS and QGIS_SERVER_PORT
```

Sample output:
```bash
QGIS Development Server listening on http://localhost:8000
CTRL+C to exit
127.0.0.1 [lun gen 20 15:16:41 2020] 5140 103ms "GET /wfs3/?MAP=/home/ale/dev/QGIS/tests/testdata/qgis_server/test_project.qgs HTTP/1.1" 200
127.0.0.1 [lun gen 20 15:16:41 2020] 3298 2ms "GET /wfs3/static/jsonFormatter.min.js HTTP/1.1" 200
127.0.0.1 [lun gen 20 15:16:41 2020] 1678 3ms "GET /wfs3/static/jsonFormatter.min.css HTTP/1.1" 200
127.0.0.1 [lun gen 20 15:16:41 2020] 1310 5ms "GET /wfs3/static/style.css HTTP/1.1" 200
127.0.0.1 [lun gen 20 15:16:43 2020] 4285 13ms "GET /wfs3/collections?MAP=/home/ale/dev/QGIS/tests/testdata/qgis_server/test_project.qgs HTTP/1.1" 200
```

### CODE REVIEW

**Comment by pblottiere** [comment]
> It can be useful when developing and testing QGIS Server projects, modules and plugins without requiring a full web server/FCGI stack.

Nice, good idea!

> The application is completely independent it does not introduce any change in the existing QGIS code, for this reason I'd like to merge it for 3.12 even if we are in feature freeze.

+1

**Comment by rldhont** [comment]
Thanks @elpaso 

@dmarteau can you take a look at this proposition.

Same comments as @pblottiere 

+1 for backport

File: src/server/qgis_mapserver.cpp [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+    } );
+
+  }
+
+  // Exit handlers
+#ifndef Q_OS_WIN
+  signal( SIGTERM, []( int ) { qApp->quit(); } );
+  signal( SIGABRT, []( int ) { qApp->quit(); } );
+  signal( SIGINT, []( int ) { qApp->quit(); } );
+  signal( SIGKILL, []( int ) { qApp->quit(); } );
```
Comment (dmarteau) [inline src/server/qgis_mapserver.cpp]: Handler is not needed  for SIGKILL because it is not catchable.

**Comment by dmarteau** [comment]
Good idea ! 

I would suggest to add command line options for setting log verbosity (i.e without having to set external environment variables).

Maybe some snippets in python could be useful too (I *really* need to take some time to wrote some of these....)

File: src/server/qgis_mapserver.cpp [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+  parser.addVersionOption();
+  parser.addPositionalArgument( QStringLiteral( "addressAndPort" ),
+                                QObject::tr( "listen to address and port (default: \"localhost:8000\")\n"
+                                    "address and port can also be specified with the environment\n"
+                                    "variables QGIS_SERVER_ADDRESS and QGIS_SERVER_PORT" ), QStringLiteral( "[address:port]" ) );
+  QCommandLineOption logLevelOption( "l", QObject::tr( "Set log level (default: 0)\n"
+                                     "0: INFO\n"
+                                     "1: WARNING\n"
+                                     "3: CRITICAL" ), "logLevel", "0" );
+  parser.addOption( logLevelOption );
```
Comment (m-kuhn) [inline src/server/qgis_mapserver.cpp]:
```suggestion
  parser.addOption( logLevelOption );  //#spellok
```

File: src/server/qgis_mapserver.cpp [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+  parser.addVersionOption();
+  parser.addPositionalArgument( QStringLiteral( "addressAndPort" ),
+                                QObject::tr( "listen to address and port (default: \"localhost:8000\")\n"
+                                    "address and port can also be specified with the environment\n"
+                                    "variables QGIS_SERVER_ADDRESS and QGIS_SERVER_PORT" ), QStringLiteral( "[address:port]" ) );
+  QCommandLineOption logLevelOption( "l", QObject::tr( "Set log level (default: 0)\n"
+                                     "0: INFO\n"
+                                     "1: WARNING\n"
+                                     "3: CRITICAL" ), "logLevel", "0" );
+  parser.addOption( logLevelOption );  //#spellok
```
Comment (3nids) [inline src/server/qgis_mapserver.cpp]:
```suggestion
  parser.addOption( logLevelOption );
```

not needed anymore

Comment (m-kuhn) [reply src/server/qgis_mapserver.cpp]: Thanks @3nids 
Sorry for not letting you do the magic first ;P

**Pull request merged**